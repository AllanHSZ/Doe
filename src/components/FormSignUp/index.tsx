import { useState } from 'react'
import { signIn } from 'next-auth/client'
import { AlertOctagon, Lock, Mail, User } from 'react-feather'

import { FieldErrors, signUpValidate } from 'utils/validations'
import { Form, FormError, FormLoading } from 'components/Form'
import Button from 'components/Button'
import Input from 'components/Input'
import axios from 'axios'

export type UserCreateProps = {
  user: string
  confirm_password: string
  password: string
  name: string
}

const FormSignUp = () => {
  const [loading, setLoading] = useState(false)
  const [formError, setFormError] = useState('')
  const [fieldError, setFieldError] = useState<FieldErrors>({})
  const [values, setValues] = useState<UserCreateProps>({
    user: '',
    password: '',
    confirm_password: '',
    name: ''
  })

  const createUser = async () => {
    setLoading(true)
    try {
      await axios.post(`${process.env.NEXT_PUBLIC_API_URL}/users`, {
        name: values.name,
        user: values.user,
        password: values.password
      })

      await signIn('credentials', {
        username: values.user,
        password: values.password,
        callbackUrl: '/'
      }).catch((error) => {
        setFormError(`${error}`)
      })
    } catch (error) {
      setFormError(`${error}`)
    } finally {
      setLoading(false)
    }
  }

  const handleInput = (field: string, value: string) => {
    setValues((s) => ({ ...s, [field]: value }))
  }

  const handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault()
    setFormError('')

    const errors = signUpValidate(values)

    if (Object.keys(errors).length) {
      setFieldError(errors)
      return
    }

    setFieldError({})
    createUser()
  }

  return (
    <Form onSubmit={handleSubmit}>
      <Input
        name="name"
        label="Nome"
        type="text"
        placeholder="Ex: JoÃ£o Silva"
        error={fieldError?.name}
        onInputChange={(v) => handleInput('name', v)}
        disabled={loading}
        icon={<User />}
      />

      <Input
        name="email"
        label="Email"
        error={fieldError?.user}
        onInputChange={(v) => handleInput('user', v)}
        disabled={loading}
        placeholder="Ex: joao.silva@email.com"
        type="email"
        icon={<Mail />}
      />

      <Input
        name="password"
        label="Senha"
        placeholder="****"
        error={fieldError?.password}
        onInputChange={(v) => handleInput('password', v)}
        disabled={loading}
        type="password"
        icon={<Lock />}
      />

      <Input
        name="confirm_password"
        label="Confirmar senha"
        error={fieldError?.confirm_password}
        onInputChange={(v) => handleInput('confirm_password', v)}
        placeholder="****"
        type="password"
        disabled={loading}
        icon={<Lock />}
      />

      {!!formError && (
        <FormError>
          <AlertOctagon /> <p>{formError}</p>
        </FormError>
      )}

      <Button type="submit" fullWidth disabled={loading}>
        {loading ? <FormLoading /> : <span>CRIAR CONTA</span>}
      </Button>
    </Form>
  )
}

export default FormSignUp
