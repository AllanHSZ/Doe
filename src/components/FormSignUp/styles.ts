import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    gap: ${theme.spacings.xxsmall};
    width: 100%;

    button {
      margin-top: ${theme.spacings.small};
      margin-right: auto;
    }
  `}
`
