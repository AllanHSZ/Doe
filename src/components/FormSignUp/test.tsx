import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'

import FormSignUp from '.'

describe('<FormSignUp />', () => {
  it('should render the form', () => {
    const { container } = renderWithTheme(<FormSignUp />)

    expect(screen.getByPlaceholderText('Ex: Jo√£o Silva')).toBeInTheDocument()
    expect(
      screen.getByPlaceholderText('Ex: joao.silva@email.com')
    ).toBeInTheDocument()
    expect(screen.queryAllByPlaceholderText('****')).toHaveLength(2)

    expect(
      screen.getByRole('button', { name: /Criar conta/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
