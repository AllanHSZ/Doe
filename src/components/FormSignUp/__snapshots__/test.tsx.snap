// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSignUp /> should render the form 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #F2F3F5;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #F2F3F5;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #04D361;
}

.c6 {
  color: #394338;
  font-family: Lato,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.4rem;
  padding: 0.8rem 0;
  padding-left: 1.6rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c6:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #F2F3F5 inset;
  -webkit-filter: none;
  filter: none;
}

.c3 {
  font-size: 1.4rem;
  color: #394338;
  margin-bottom: 0.4rem;
  cursor: pointer;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 2.2rem;
  color: #8F8F8F;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c5 > svg {
  width: 100%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-family: Lato,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  border: 0;
  cursor: pointer;
  border-radius: 0.6rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #FAFAFA;
  background: #04D361;
  height: 4rem;
  font-size: 1.4rem;
  font-weight: 700;
  padding: 0.8rem 3.2rem;
  width: 100%;
}

.c8:hover {
  background: #04AB4F;
}

.c0 {
  width: 100%;
}

.c0 .c1 {
  margin: 1.2rem 0;
}

.c0 .c7 {
  margin-top: 1.6rem;
}

<form
  class="c0"
>
  <div
    class="c1 c2"
  >
    <label
      class="c3"
      for="name"
    >
      Nome
    </label>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 21v-2a4 4 0 0 0-4-4H8a4 4 0 0 0-4 4v2"
          />
          <circle
            cx="12"
            cy="7"
            r="4"
          />
        </svg>
      </div>
      <input
        class="c6"
        id="name"
        name="name"
        placeholder="Ex: JoÃ£o Silva"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="c1 c2"
  >
    <label
      class="c3"
      for="email"
    >
      Email
    </label>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
          />
          <polyline
            points="22,6 12,13 2,6"
          />
        </svg>
      </div>
      <input
        class="c6"
        id="email"
        name="email"
        placeholder="Ex: joao.silva@email.com"
        type="email"
        value=""
      />
    </div>
  </div>
  <div
    class="c1 c2"
  >
    <label
      class="c3"
      for="password"
    >
      Senha
    </label>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="11"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="11"
          />
          <path
            d="M7 11V7a5 5 0 0 1 10 0v4"
          />
        </svg>
      </div>
      <input
        class="c6"
        id="password"
        name="password"
        placeholder="****"
        type="password"
        value=""
      />
    </div>
  </div>
  <div
    class="c1 c2"
  >
    <label
      class="c3"
      for="confirm_password"
    >
      Confirmar senha
    </label>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="11"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="11"
          />
          <path
            d="M7 11V7a5 5 0 0 1 10 0v4"
          />
        </svg>
      </div>
      <input
        class="c6"
        id="confirm_password"
        name="confirm_password"
        placeholder="****"
        type="password"
        value=""
      />
    </div>
  </div>
  <button
    class="c7 c8"
    type="submit"
  >
    <span>
      <span>
        CRIAR CONTA
      </span>
    </span>
  </button>
</form>
`
